src/Interpreters/Streaming/HashJoin.cpp

```diff
@@ -1792,11 +1829,88 @@ Block HashJoin::insertRightBlockAndJoin(Block & right_block)
 
     std::scoped_lock lock(left_data.buffered_data->mutex, right_data.buffered_data->mutex);
 
-    /// If this is a retract block which contains _tp_delta with -1 as it value
-    if (auto joined_retracted_block = eraseExistingKeysAndRetractJoin<false>(right_block); joined_retracted_block)
-        return std::move(*joined_retracted_block);
+    if (emitChangeLog())
+    {
+        /// If this is a retract block which contains _tp_delta with -1 as it value
+        if (auto joined_retracted_block = eraseExistingKeysAndRetractJoin<false>(right_block); joined_retracted_block)
+            return std::move(*joined_retracted_block);
+    }
+    else
+    {
+        /// Just erase the existing keys from hash map on right side, no need retract join
+        if (isRetractBlock(right_block, *right_data.join_stream_desc))
+        {
+            eraseExistingKeys<false>(right_block, right_data);
+            right_block.clear();
+            return {};
+        }
+    }
 
-    doInsertBlock<false>(right_block, right_data.buffered_data->getCurrentHashBlocksPtr());
+    /// For left join, if right data have new join keys, we need delete prev joined result (join key + left_cols + right_cols<empty>),
```

Comment:
`/// For left join, if right data have new join keys, we need delete prev left joined result (join key + left_cols + empty right_cols),`